accepted/00000-auto-generate-bootstrap-repositories.md

```diff
@@ -0,0 +1,71 @@
+[motivation]: #motivation
+
+To be able to register a system to Uyuni Server you need to have the software installed which provide the required tooling.
+
+A bootstrap repository provide the software for the supported operating systems with its dependencies. The content for the
+bootstrap repositories come from the normal channels which must be synchronized and updated on regular base.
+
+Users very often do not know this, or forget to create them when they synchronize a new product. Also the bootstrap repository
+is only updated when the user execute the generator script again.
+
+This RFC should describe how to change the current tools to automatically 
+execute the generation of bootstrap repositories after
+a new product was added and after a channel changed when a maintenance update was synced to Uyuni Server.
+
+# Detailed design
+[design]: #detailed-design
+
+The currently exiting tool `mgr-create-bootstrap-repo` list all available bootstrap repositories it could generate.
+This list depends on the products the user has synchronized.
+
+Step 1: provide an option to enable an automatic mode which iterate over all available bootstrap repositories and generate them
+when needed. As we have sometimes have multiple options for the same local "path" we should iterate over the paths to minimize
+the number of runs.
+
+Step 2: add a log file to `mgr-create-bootstrap-repo`. Currently it has no log file. But when this tool is running in the background
+we need the possibility to debug problems. All messages should go to a log file.
+
+Step 3: let `mgr-create-bootstrap-repo` use a configuration file. All commandline options should be possible to set via
+configuration file. This allows to influence the behavior even in automatic mode when it is not possible to change the commandline
+options.
+
+To trigger the re-generation we implement the following algorithm:
+
+- Create a taskomatic job running every 15 minutes and call `mgr-create-bootstrap-repo --auto`
+- It calculate all channels which are needed to create the bootstrap repository for every available OS
+- It compare the `last_modified` timestamps of the channels (DB) with the `modified date` of the main metadata file
+  (repomd.xml or Release)
+- if all `last_modified` timestamps are newer then the timestamp of the main metadata file, re-generate the bootstrap repository
+- if any `last_modified` timestamps are newer then the timestamp of the main metadata file and the `modified date` of the main metadata
+  file is older than 26 hours, re-generate the bootstrap repository. This is for the case a channel is not synced. It can happen on
+  custom repositories needed for bootstrap repositories.
+
+Provide a `--force` option to enforce recreation of all bootstrap repositories without checking if needed. Mostly for manual usage.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+We maybe 15 minutes late with the bootstrap repository or maybe longer when a channel failed to sync or does not sync at all.
```

Comment:
What happens if the bootstrap repository creation takes more than 15 min? Will another run of the algorithm and `mgr-create-bootstrap-repo be started? How do we protect against that?

Can we do this based on actual packages being new/updated, and then run the check for the flag every 1 min, so that if there is no new/updated packages nothing will be done (so check finishes early) and if there are new/updated packages the bootstrap repository is immediately generated?

Reply:
Running multiple `mgr-create-bootstrap-repo` at the same time can be prevented with taskomatic or locking.
spacewalk-repo-sync does the same. We can copy that mechanism if we want. I will add the locking to the RFC.

About the second question: It is very hard as we have the data only inside of the `mgr-create-bootstrap-repo` tool.
This is the reason why I want to call the tool, and the tool check if there is something to do or not.
In case there is nothing to do, it should exit as soon as possible.
We can talk about the interval when we know how long it would take to find out that there is nothing to do.
But doing things every minute will add load on the server. And when I add a new product we have to wait anyway for all required channels to finish syncing. Depending on the product it takes between 30 Minutes and multiple hours.
So I think 15 minutes is not that much.

Reply:
_Running multiple mgr-create-bootstrap-repo at the same time can be prevented with taskomatic or locking.
spacewalk-repo-sync does the same. We can copy that mechanism if we want. I will add the locking to the RFC._

I don't know the internals of mgr-create-bootstrap-repo but problem should not be running multiple mgr-create-bootstrap-repo at the same time but multiple mgr-create-bootstrap-repo at the same time *for the same product*, which is what may clash.

Reply:
Right.